### PR TITLE
Implement KOKKOS_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND

### PR DESCRIPTION
**Pull request by masterleinad**
Fixing thread-safety issues in https://github.com/kokkos/kokkos/pull/7080 and https://github.com/kokkos/kokkos/pull/6151 causes issues in (wrong) View of View code that doesn't manually call the destructor on the interior Views.
This pull request introduces a CMake option that allows serializing calls to the View's value type's destructor working around the issue.

### CODE REVIEW

File: cmake/kokkos_enable_options.cmake [hunk 1]
```diff
@@ -76,6 +76,7 @@ KOKKOS_ENABLE_OPTION(IMPL_HIP_UNIFIED_MEMORY OFF "Whether to leverage unified me
 # This option will go away eventually, but allows fallback to old implementation when needed.
 KOKKOS_ENABLE_OPTION(DESUL_ATOMICS_EXTERNAL OFF "Whether to use an external desul installation")
 KOKKOS_ENABLE_OPTION(ATOMICS_BYPASS OFF "**NOT RECOMMENDED** Whether to make atomics non-atomic for non-threaded MPI-only use cases")
+KOKKOS_ENABLE_OPTION(IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND OFF "**NOT RECOMMENDED** Whether to enable a workaround for invalid View of Views code.")
```
Comment (dalg24) [inline cmake/kokkos_enable_options.cmake:79]:
```suggestion
KOKKOS_ENABLE_OPTION(IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND OFF "Whether to enable a workaround for invalid use of View of Views that causes program hang on destruction.")
mark_as_advanced(Kokkos_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND)
```

Comment (dalg24) [reply cmake/kokkos_enable_options.cmake:79]: I take it as you didn't like my `_GRACE_PERIOD` suggestion?

Comment (masterleinad) [reply cmake/kokkos_enable_options.cmake:79]: Yeah, I wasn't quite sure about that. I'm not set on the current name, though.

**Comment by dalg24** [comment]
(got rid of the trailing new line that tripped the format check)

File: cmake/KokkosCore_config.h.in [hunk 2]
```diff
@@ -128,3 +128,4 @@
 #cmakedefine KOKKOS_ARCH_NAVI1100 // deprecated
 
 #cmakedefine KOKKOS_IMPL_32BIT
+#cmakedefine KOKKOS_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND
```
Comment (dalg24) [inline cmake/KokkosCore_config.h.in:131]: It is misplaced.  Please put it with the other ENABLE_ options

File: core/unit_test/TestViewOfViews.hpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    vov(0, 0) = a;
+    vov(1, 0) = a;
+    vov(0, 1) = b;
+#ifndef KOKKOS_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND
+    vov(0, 0) = V();
+    vov(1, 0) = V();
+    vov(0, 1) = V();
+#endif
+  }
+  {  // using placement new to construct the innver objects and explicitly
```
Comment (dalg24) [inline core/unit_test/TestViewOfViews.hpp:50]:
```suggestion
  {  // using placement new to construct the inner objects and explicitly
```

**Comment by dalg24** [comment]
The motivation for adding this IMPL_ option is we discovered that misuse is quite widespread downstream

**Review by masterleinad — COMMENTED** [review]
The test looks good to me.

File: cmake/kokkos_enable_options.cmake [hunk 4]
```diff
@@ -76,6 +76,8 @@ KOKKOS_ENABLE_OPTION(IMPL_HIP_UNIFIED_MEMORY OFF "Whether to leverage unified me
 # This option will go away eventually, but allows fallback to old implementation when needed.
 KOKKOS_ENABLE_OPTION(DESUL_ATOMICS_EXTERNAL OFF "Whether to use an external desul installation")
 KOKKOS_ENABLE_OPTION(ATOMICS_BYPASS OFF "**NOT RECOMMENDED** Whether to make atomics non-atomic for non-threaded MPI-only use cases")
+KOKKOS_ENABLE_OPTION(IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND OFF "Whether to enable a workaround for invalid use of View of Views that causes program hang on destruction.")
+mark_as_advanced(Kokkos_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND)
```
Comment (ldh4) [inline cmake/kokkos_enable_options.cmake:80]: What does this line do?

Comment (dalg24) [reply cmake/kokkos_enable_options.cmake:80]: It "hides" the option.

https://cmake.org/cmake/help/latest/command/mark_as_advanced.html

**Review by nmm0 — APPROVED** [review]
I'm fine with this as long as we eventually remove this option

File: core/unit_test/TestViewOfViews.hpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    new (&vov(0, 0)) V(a);
+    new (&vov(1, 0)) V(a);
+    new (&vov(0, 1)) V(b);
+#ifndef KOKKOS_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND
+    vov(0, 0).~V();
+    vov(1, 0).~V();
+    vov(0, 1).~V();
+#else
+    // leaks memory
+#endif
```
Comment (ldh4) [inline core/unit_test/TestViewOfViews.hpp:64]: I'd like to suggest adding in a short note here that this leak only happens for threaded backend.

**Review by ldh4 — APPROVED** [review]
I'd like to

**Comment by dalg24** [comment]
> I'm fine with this as long as we eventually remove this option

This is really a contingency plan so that the 4.4 release is not held back.  We introduce a hidden IMPL_ option that we intend to remove potentially even at the next minor release.

File: core/unit_test/TestViewOfViews.hpp (hunk 5 above)
Comment (dalg24) [reply core/unit_test/TestViewOfViews.hpp:64]: No the leak would happen for any backend.

File: core/src/View/Kokkos_ViewAlloc.hpp [hunk 6]
```diff
@@ -182,7 +182,15 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
   void construct_shared_allocation() { construct_dispatch(); }
 
   void destroy_shared_allocation() {
-    parallel_for_implementation<DestroyTag>();
+#ifdef KOKKOS_ENABLE_IMPL_VIEW_OF_VIEWS_DESTRUCTOR_PRECONDITION_VIOLATION_WORKAROUND
+    if constexpr (std::is_same_v<typename ExecSpace::memory_space,
+                                 Kokkos::HostSpace>)
+      for (int64_t i = 0; i < n; ++i) (ptr + i)->~ValueType();
```
Comment (dalg24) [inline core/src/View/Kokkos_ViewAlloc.hpp:188]: Seeing warnings on my local build
```
core/src/View/Kokkos_ViewAlloc.hpp:188:29: warning: comparison of integers of different signs: 'int64_t' (aka 'long long') and 'size_t' (aka 'unsigned long') [-Wsign-compare]
      for (int64_t i = 0; i < n; ++i) (ptr + i)->~ValueType();

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
(Couldn't think of anything better to re-trigger the CI.  Sorry about the noise...)

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
Retest this please

**Pull request merged**